### PR TITLE
Adds Python 3 support

### PR DESCRIPTION
**Pull request by Garethp**
Adds Python 3 support and opens some functionality as an open API for other plugins to be able to use

### CODE REVIEW

**Comment by BlackIkeEagle** [comment]
@Garethp could you update the .travis.yml file with python3 versions so we can see if there are any issues related to that?

  - "3.2"
  - "3.3"
  - "3.4"
  - "3.5"
  - "3.6"

**Comment by Garethp** [comment]
I've done the update, but I haven't updated the tests, so the tests might fail. I'll update those later

**Comment by BlackIkeEagle** [comment]
@Garethp great work anyway, I have to find some time to test some stuff :)

**Comment by Garethp** [comment]
Also, I'm not sure if you're interested, but in my fork I've reworked the layout a little bit to use the space better, similar to PHPStorm (Debugger on the bottom, status window being reduced to one line for Status and Keys (It looks like this `  ▌▌  [<F5> Start] [<F6> Stop] [:help Vdebug]`, with expandable for connection info, the status and stack window being reduced to 60 column width to give more space for the watcher

**Comment by BlackIkeEagle** [comment]
@Garethp I'm interested in every improvement :+1:

**Comment by Garethp** [comment]
One thing to note is that I had to remove a lot of base64 stuff in some places, as for some reason the data wasn't encoded either way. Not sure if that'll screw up other debuggers. Do you know much about that?

**Comment by BlackIkeEagle** [comment]
@Garethp I test with php xdebug so I have no clue if there will be issues for other languages

**Comment by Garethp** [comment]
One of the features I'm adding is pressing `<Leader>E` in normal mode to be able to type in any expression manually, instead of needing it to be in the current buffer to evaluate it. Quick Question: In PHPStorm, when you do evaluate it actually changes the values of those variables in the current running script, but I'm not sure how? All the commands in dbpg that look like they'd do that seem to be unimplemented in xdebug 2.5.0

**Comment by BlackIkeEagle** [comment]
@Garethp I'm not sure what you mean?
> In PHPStorm, when you do evaluate it actually changes the values of those variables in the current running script

so you say if $a = 3 and you evaluate $a++ in PHPStorm $a will become 4?

**Comment by BlackIkeEagle** [comment]
@Garethp Btw, I'm testing some other stuff currently based on @joonty 's initial work in the version2 branch now in the v2-integration branch. I'll try to add your python3 changes in that branch lateron

**Comment by Garethp** [comment]
Exactly. Take the following

```php
$a = 1;
$b = 2;
$c = $a * $b;
echo $c;
```

If you break before the third line, evaluate `$a = 5;` then continue, `$c` will end up equalling `10`, not `2`

**Comment by BlackIkeEagle** [comment]
@Garethp I'm going to have to do some work to get this to work with the v2-integration branch since there were a lot of changes there

**Comment by jeffrichards** [comment]
Just a comment that with debian stretch release the vim-nox plugin appears to no longer have support for the python 2, and as such the plugin doesn't work unless it can use the python 3.

**Comment by BlackIkeEagle** [comment]
@jeffrichards I'm very much aware of the ubuntu and debian related issues. But we have to get everything working properly. I can't get this update to merge properly in the v2-integration branch and that branch also has some other issues to fix

File: plugin/vdebug.vim [hunk 1]
```diff
@@ -18,9 +18,9 @@
 "=============================================================================
 " }}}
 
-" Do not source this script when python is not compiled in.
+" Do not source this script when python3 is not compiled in.
 if !has("python")
```
Comment (lucc) [inline plugin/vdebug.vim]: This should be "python3"

File: plugin/vdebug.vim [hunk 2]
```diff
@@ -18,9 +18,9 @@
...
-    finish
+    "finish
 endif
```
Comment (lucc) [inline plugin/vdebug.vim]: Why is this commented out?

Comment (Garethp) [inline plugin/vdebug.vim]: Will change this, and the python check

**Comment by BlackIkeEagle** [comment]
@Garethp would you want to check how much effort would be required to get the v2-integration branch to python3? I would like to get this more async and the python3 stuff out together. @joonty did some refactoring in v2 which makes this nice python3 update not so straight forward to merge there.

**Comment by Garethp** [comment]
I'll see if I can find the time over the next week or so

**Comment by BlackIkeEagle** [comment]
@Garethp thank you very much, @lucc already opened some PR's versus v2-integration, I'll merge them soon

**Comment by l0rb** [comment]
There is now a bounty on this. https://www.bountysource.com/issues/46922753-support-for-python-3

**Comment by BlackIkeEagle** [comment]
@l0rb fortunately @lucc mentioned the v2-integration branch where there is continued development but still some issues to be resolved before thinking about py3

**Comment by BlackIkeEagle** [comment]
thanks for the intial work, the python3 support is now in v2-integration

**Pull request closed without merging**